docs/source/notes/distributed_training.rst

```diff
@@ -135,5 +135,91 @@ from the trainer.
     Not all algorithms have been tested with Deepspeed, please proceed with
     caution.
 
+
+FullyShardedDataParallel (FSDP)
+---------
+
+Composer integrates Pytorch's `FullyShardedDataParallel <https://pytorch.org/docs/stable/fsdp.html>`__ engine with some syntactic sugar to make it easy to write custom models that work with Composer + FSDP.
+
+At a high level, when you use the Composer Trainer, you must pass it a :mod:`ComposerModel` like `ComposerGPT <https://github.com/mosaicml/benchmarks/blob/6972fe3000d5a5480d8757ff710965514155e8db/llm/llm/gpt.py#L178>`__ that defines certain functions like :code:`forward`, :code:`eval_forward`, :code:`loss`, etc. that are called during the training loop.
+
+Inside that :mod:`ComposerModel` you may have one or many submodules, such as a :code:`.model` or :code:`.language_model` or :code:`.classifier` that is the actual :mod:`torch.nn.Module` that you will be deploying at inference time. In our case, this is the `GPT <https://github.com/mosaicml/benchmarks/blob/6972fe3000d5a5480d8757ff710965514155e8db/llm/llm/gpt.py#L102>`__ module that we build and attach :mod:`ComposerGPT.model`.
+
+When you provide an :code:`fsdp_config={...}` dictionary to the Composer Trainer, then on :code:`__init__`, the Trainer will attempt to wrap **each of the submodules** of your :mod:`ComposerModel` with an FSDP auto wrap policy. This wrapping is recursive, so not only is `GPT` wrapped, but all submodules of `GPT` may/may not be wrapped too. See the `FSDP documentation <https://pytorch.org/docs/stable/fsdp.html>`__ for more details on how auto wrap policies work.
+
+The full spec and defaults for Composer's `fsdp_config` is here:
+
+.. code:: python
+
+    fsdp_config = {
+      'sharding_strategy': str = 'FULL_SHARD' | 'SHARD_GRAD_OP' | 'NO_SHARD', # Default: 'FULL_SHARD'
+      'min_params': float # Default: 1e8
+      'cpu_offload': bool = True | False, # Default: False, cpu_offload not supported yet
+      'mixed_precision': str = 'full' | 'default' | 'pure', # Default: 'default'
+      # Note: you can explictly provide a dictionary too
+      # 'mixed_precision': dict = {
+      #   'param_dtype': 'fp32' | 'fp16' | 'bf16',
+      #   'reduce_dtype': 'fp32' | 'fp16' | 'bf16',
+      #   'buffer_dtype': 'fp32' | 'fp16' | 'bf16',
+      # },
+      'backward_prefetch': str = 'BACKWARD_PRE' | 'BACKWARD_POST' | 'NONE', # Default: 'BACKWARD_POST'
+      'activation_checkpointing': bool = True | False, # Default: False
+      'activation_cpu_offload': bool = True | False, # Default: False
+    }
```

Comment:
Need to add `verbose: bool = True | False,` that's on me I forgot about it previously.